### ISSUE TITLE
Mutable variable is considered immutable

### ISSUE THREAD

**Issue by Braqzen** [FuelLabs/sway#5803]
Forc: `0.52.0`

```sway
contract;

pub struct Balance {
    base: u64,
}

impl Balance {
    pub fn credit(ref mut self, amount: u64) {
        self.base += amount;
    }
}

abi MyContract {
    fn test_function();
}

impl MyContract for Contract {
    fn test_function() {
        let x = Some(Balance { base: 0 });
        match x {
            Some(mut balance) => balance.credit(5),
            None => revert(0),
        };
    }
}
```

![1](https://github.com/FuelLabs/sway/assets/103777923/3d037bfe-3273-4387-b343-fa0d0b0b70d8)

**Comment by IGI-111** [comment]
`x` is immutable (as in `let x`), therefore elements of it are also immutable.

However this also fails:

```sway
contract;

pub struct Balance {
    base: u64,
}

impl Balance {
    pub fn credit(ref mut self, amount: u64) {
        self.base += amount;
    }
}

abi MyContract {
    fn test_function();
}

impl MyContract for Contract {
    fn test_function() {
        let mut x = Some(Balance { base: 0 }); // NOTE: added mut
        match x {
            Some(mut balance) => balance.credit(5),
            None => revert(0),
        };
    }
}
```

But more importantly it doesn't have the right semantics. What you want isn't `mut balance` but `ref mut balance`, which is unsupported as of now.

@ironcev what do you think about this?

**Comment by ironcev** [comment]
These features, `mut balance` and `ref mut balance`, are currently not implemented. When I started working on references I noticed that and left a TODO in code to implement it. `ref` and `mut` in patterns exist on the AST level but are ignored in type checking. Actually, if you write `ref mut` you will get an error that `ref` patterns are not supported. Having only `mut` will not emit any error but will be completely ignored in type checking which justifies the error message that `balance` is not mutable.

To avoid the confusion until we implement the feature, we should emit the not supported error even if only `mut` is used. I'll do that.
